File: assets/js/hooks/dimension_drift.js
import * as d3 from "d3"
import { DIM_COLORS } from "./constants"

/** Small-multiples sparklines — 90-day rolling average per dimension, showing long-term drift. */
const MARGIN = { top: 6, right: 15, bottom: 24, left: 72 }
const WIDTH = 800
const ROW_H = 44
const GAP = 8
const parseDate = d3.timeParse("%Y-%m-%d")

export const DimensionDrift = {
  mounted() {
    this.handleEvent("update-drift", (data) => {
      this.data = data
      this.render()
    })
  },

  render() {
    const { dimensions } = this.data
    if (!dimensions || dimensions.length === 0) return

    this.el.innerHTML = ""

    const totalH = MARGIN.top + MARGIN.bottom + dimensions.length * (ROW_H + GAP) - GAP
    const innerW = WIDTH - MARGIN.left - MARGIN.right

    const svg = d3.select(this.el)
      .append("svg")
      .attr("viewBox", `0 0 ${WIDTH} ${totalH}`)
      .attr("preserveAspectRatio", "xMidYMid meet")
      .style("width", "100%")

    // Shared x scale
    const allDates = dimensions[0].series.map(d => parseDate(d.date))
    const x = d3.scaleTime()
      .domain(d3.extent(allDates))
      .range([0, innerW])

    dimensions.forEach((dim, i) => {
      const rowY = MARGIN.top + i * (ROW_H + GAP)
      const color = DIM_COLORS[dim.dimension] || "#888"

      const g = svg.append("g")
        .attr("transform", `translate(${MARGIN.left}, ${rowY})`)

      // Per-row y scale
      const values = dim.series.map(d => d.value)
      const yMin = d3.min(values)
      const yMax = d3.max(values)
      const yPad = (yMax - yMin) * 0.15 || 0.5
      const y = d3.scaleLinear()
        .domain([yMin - yPad, yMax + yPad])
        .range([ROW_H, 0])

      // Subtle background
      g.append("rect")
        .attr("width", innerW)
        .attr("height", ROW_H)
        .attr("fill", "rgba(255,255,255,0.015)")
        .attr("rx", 3)

      // Overall mean reference line
      const mean = d3.mean(values)
      g.append("line")
        .attr("x1", 0)
        .attr("x2", innerW)
        .attr("y1", y(mean))
        .attr("y2", y(mean))
        .attr("stroke", "rgba(255,255,255,0.08)")
        .attr("stroke-width", 0.5)
        .attr("stroke-dasharray", "3,3")

      // Rolling average line
      const line = d3.line()
        .x(d => x(parseDate(d.date)))
        .y(d => y(d.value))
        .curve(d3.curveBasis)

      g.append("path")
        .datum(dim.series)
        .attr("d", line)
        .attr("fill", "none")
        .attr("stroke", color)
        .attr("stroke-width", 1.5)
        .attr("stroke-opacity", 0.8)

      // Dimension label
      svg.append("text")
        .attr("x", MARGIN.left - 8)
        .attr("y", rowY + ROW_H / 2)
        .attr("text-anchor", "end")
        .attr("dominant-baseline", "central")
        .attr("fill", color)
        .attr("font-size", "10px")
        .attr("font-weight", "600")
        .text(dim.dimension)

      // Scale hints: first and last value
      const first = values[0]
      const last = values[values.length - 1]

      g.append("text")
        .attr("x", -2)
        .attr("y", y(first))
        .attr("text-anchor", "end")
        .attr("dominant-baseline", "central")
        .attr("fill", "#444")
        .attr("font-size", "8px")
        .text(first.toFixed(1))

      g.append("text")
        .attr("x", innerW + 4)
        .attr("y", y(last))
        .attr("dominant-baseline", "central")
        .attr("fill", "#555")
        .attr("font-size", "8px")
        .text(last.toFixed(1))
    })

    // Shared x axis at bottom
    const axisY = MARGIN.top + dimensions.length * (ROW_H + GAP) - GAP
    const xAxisG = svg.append("g")
      .attr("transform", `translate(${MARGIN.left}, ${axisY})`)
      .call(d3.axisBottom(x).ticks(d3.timeYear.every(1)).tickFormat(d3.timeFormat("%Y")).tickSize(3))
    xAxisG.select(".domain").attr("stroke", "#444")
    xAxisG.selectAll(".tick line").attr("stroke", "#444")
    xAxisG.selectAll(".tick text").attr("fill", "#666").attr("font-size", "10px")
  }
}

File: lib/fugue_web/live/mood_live/dimension_drift.ex
defmodule FugueWeb.MoodLive.DimensionDrift do
  @moduledoc """
  Small-multiples sparklines — 90-day rolling average per dimension, showing
  long-term drift. Replaces the former `DimensionDrift` JS hook.
  """

  use Phoenix.Component

  alias FugueWeb.MoodLive.SvgMath

  @width 800
  @m_top 6
  @m_right 15
  @m_bottom 24
  @m_left 72
  @row_h 44
  @gap 8

  @dim_colors %{
    "sleep" => "#42c8e6",
    "anxiety" => "#e44dbc",
    "sensitivity" => "#a86ee6",
    "outlook" => "#6ee64d",
    "speed" => "#e6a542"
  }

  attr :dimensions, :list, default: []

  def drift(assigns) do
    dims = assigns.dimensions
    inner_w = @width - @m_left - @m_right

    {rows, x_ticks, total_h, axis_y} =
      if dims == [] do
        {[], [], 100, 0}
      else
        dates = Enum.map(List.first(dims).series, & &1.date)
        {min_d, max_d} = date_range(dates)
        span = max(Date.diff(max_d, min_d), 1)

        x_fn = fn d -> Date.diff(Date.from_iso8601!(d), min_d) / span * inner_w end

        rows =
          Enum.with_index(dims)
          |> Enum.map(fn {dim, i} ->
            color = Map.get(@dim_colors, dim.dimension, "#888")
            values = Enum.map(dim.series, & &1.value)
            y_min = Enum.min(values)
            y_max = Enum.max(values)
            y_pad = max((y_max - y_min) * 0.15, 0.5)
            domain_lo = y_min - y_pad
            domain_hi = y_max + y_pad
            domain_span = max(domain_hi - domain_lo, 1.0e-9)

            y_fn = fn v ->
              @row_h - (v - domain_lo) / domain_span * @row_h
            end

            points = Enum.map(dim.series, fn p -> {x_fn.(p.date), y_fn.(p.value)} end)
            mean = Enum.sum(values) / length(values)
            first_val = List.first(values)
            last_val = List.last(values)

            %{
              dimension: dim.dimension,
              color: color,
              row_y: @m_top + i * (@row_h + @gap),
              path_d: SvgMath.basis_path(points),
              mean_y: y_fn.(mean),
              first_val: first_val,
              first_y: y_fn.(first_val),
              last_val: last_val,
              last_y: y_fn.(last_val),
              label_y: @m_top + i * (@row_h + @gap) + @row_h / 2
            }
          end)

        total_h = @m_top + @m_bottom + length(dims) * (@row_h + @gap) - @gap
        axis_y = @m_top + length(dims) * (@row_h + @gap) - @gap

        x_ticks =
          for y <- min_d.year..max_d.year,
              jan1 = Date.new!(y, 1, 1),
              Date.compare(jan1, min_d) != :lt and Date.compare(jan1, max_d) != :gt do
            %{x: x_fn.(Date.to_iso8601(jan1)), label: Integer.to_string(y)}
          end

        {rows, x_ticks, total_h, axis_y}
      end

    assigns =
      assign(assigns,
        rows: rows,
        x_ticks: x_ticks,
        svg_width: @width,
        svg_height: total_h,
        axis_y: axis_y,
        inner_w: inner_w,
        row_h: @row_h,
        m_left: @m_left,
        label_x: @m_left - 8,
        axis_transform: "translate(#{@m_left},#{axis_y})"
      )

    ~H"""
    <div id="dimension-drift" style="width: 100%;">
      <svg
        viewBox={"0 0 #{@svg_width} #{@svg_height}"}
        preserveAspectRatio="xMidYMid meet"
        style="width: 100%;"
      >
        <%= for row <- @rows do %>
          <g transform={"translate(#{@m_left},#{SvgMath.fmt(row.row_y)})"}>
            <rect width={@inner_w} height={@row_h} fill="rgba(255,255,255,0.015)" rx="3" />
            <line
              x1="0"
              x2={@inner_w}
              y1={SvgMath.fmt(row.mean_y)}
              y2={SvgMath.fmt(row.mean_y)}
              stroke="rgba(255,255,255,0.08)"
              stroke-width="0.5"
              stroke-dasharray="3,3"
            />
            <path
              d={row.path_d}
              fill="none"
              stroke={row.color}
              stroke-width="1.5"
              stroke-opacity="0.8"
            />
            <text
              x="-2"
              y={SvgMath.fmt(row.first_y)}
              text-anchor="end"
              dominant-baseline="central"
              fill="#444"
              font-size="8px"
            >
              {num(row.first_val)}
            </text>
            <text
              x={@inner_w + 4}
              y={SvgMath.fmt(row.last_y)}
              dominant-baseline="central"
              fill="#555"
              font-size="8px"
            >
              {num(row.last_val)}
            </text>
          </g>

          <text
            x={@label_x}
            y={SvgMath.fmt(row.label_y)}
            text-anchor="end"
            dominant-baseline="central"
            fill={row.color}
            font-size="10px"
            font-weight="600"
          >
            {row.dimension}
          </text>
        <% end %>

        <g transform={@axis_transform}>
          <line x1="0" y1="0" x2={@inner_w} y2="0" stroke="#444" />
          <%= for t <- @x_ticks do %>
            <line x1={SvgMath.fmt(t.x)} y1="0" x2={SvgMath.fmt(t.x)} y2="3" stroke="#444" />
            <text
              x={SvgMath.fmt(t.x)}
              y="16"
              text-anchor="middle"
              fill="#666"
              font-size="10px"
            >
              {t.label}
            </text>
          <% end %>
        </g>
      </svg>
    </div>
    """
  end

  defp num(v) when is_float(v), do: :erlang.float_to_binary(v, decimals: 1)
  defp num(v), do: to_string(v)

  defp date_range([first | _] = dates) do
    parsed = Enum.map(dates, &Date.from_iso8601!/1)
    {Enum.min(parsed, Date), Enum.max(parsed, Date)}
  rescue
    _ -> {Date.from_iso8601!(first), Date.from_iso8601!(first)}
  end
end
